plugin.rb

```diff
@@ -64,7 +66,38 @@ def render_bcc(status:)
       return render_bcc(status: false) { |result| result.add_error(post.errors[:raw]) }
     end
 
-    Jobs.enqueue(:bcc_post, user_id: current_user.id, create_params: @manager_params)
+    # Queue up jobs in batches so that when sending hundreds or thousands of emails we
+    # can take advantage of multiple workers
+    batch_start = 0
+    batch_end = DiscourseBCC::BATCH_SIZE - 1
+    @manager_params.except(:target_users, :target_group_names, :target_emails)
+    usernames = usernames.to_a
+    emails = emails.to_a
+    while batch_start <= usernames.size && usernames.size > 0
+      Jobs.enqueue(
+        :bcc_post,
+        user_id: current_user.id,
+        create_params: @manager_params,
+        targets_key: 'target_usernames',
+        targets: usernames[batch_start..batch_end]
+      )
+      batch_start += DiscourseBCC::BATCH_SIZE
+      batch_end += DiscourseBCC::BATCH_SIZE
+    end
+
+    batch_start = 0
+    batch_end = DiscourseBCC::BATCH_SIZE - 1
+    while batch_start <= emails.size && emails.size > 0
+      Jobs.enqueue(
+        :bcc_post,
+        user_id: current_user.id,
+        create_params: @manager_params,
+        targets_key: 'target_emails',
+        targets: emails[batch_start..batch_end]
+      )
+      batch_start += DiscourseBCC::BATCH_SIZE
+      batch_end += DiscourseBCC::BATCH_SIZE
```

Comment:
The variables seem to be the while condition limit, `targets_key` and `targets`. Just wondering if it's nicer to create a function for the looping, so if we make changes to one loop, the other loop gets it for sure too.

Reply:
Yes I thought about creating a function for that but we are patching the class and didn't think there was a clean way to do that?

Reply:
Additionally, do we have a test that has > 20 usernames / emails to check that the job gets enqueued?

Reply:
Nope, but I can add that 👍.

Reply:
> we are patching the class

Aha, I had assumed that invoking a `def batch_enqueue` within `add_to_class` would work. Perhaps with the tests in place we can confirm that it won't work.

Reply:
@nattsw lol there is an example of adding a protected method in this very plugin, `render_bcc`. I'll add another method for the loop.